replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,354 @@
+    if (!processing) {
+      this.running = false;
+      return true;
+    }
+    return false;
+  }
+
+  private boolean supported(Task task) {
+    if (!OperationType.HARVEST.equals(task.getOperation())) {
+      return false;
+    }
+
+    List<MetadataInfo> infos = task.metadatas().collect(Collectors.toList());
+    if (infos.size() > 1) {
+      LOGGER.warn("More than one metadata is currently not supported by this worker");
+      return false;
+    }
+
+    return (infos.stream().filter(DdfMetadataInfo.class::isInstance).count() == 1L);
+  }
+
+  @SuppressWarnings(
+      "squid:S3655" /* Optional.get() will return an info because we check it exists in the supported method */)
+  private void handle(Task task) throws InterruptedException {
+    NodeAdapter remote = getDestinationAdapter(task);
+    if (remote == null) {
+      task.unlock();
+      return;
+    }
+
+    try {
+      if (!adaptersAvailable(task, remote)) {
+        return;
+      }
+
+      DdfMetadataInfo info =
+          task.metadatas()
+              .filter(DdfMetadataInfo.class::isInstance)
+              .map(DdfMetadataInfo.class::cast)
+              .findFirst()
+              .get(); // should never end up being empty since it was filtered above
+
+      final Metadata metadata =
+          new MetadataImpl(
+              info.getData(), info.getDataClass(), task.getId(), Date.from(info.getLastModified()));
+
+      final OperationType type = task.getOperation();
+      if (type == OperationType.HARVEST) {
+        doCreate(task, metadata, remote, local);
+      } else {
+        // should never hit this since deserialization will return UNKNOWN
+        task.unlock();
+      }
+    } finally {
+      try {
+        remote.close();
+      } catch (IOException e) {
+        LOGGER.debug("Failed to close destination adapter", e);
+      }
+    }
+  }
+
+  @Nullable
+  private NodeAdapter getDestinationAdapter(Task task) {
+    Site site;
+    final String siteId = task.getQueue().getSite();
+    try {
+      site = siteManager.get(siteId);
+    } catch (NotFoundException e) {
+      LOGGER.debug("Unable to find site {} for queue", siteId, e);
+      return null;
+    } catch (ReplicationPersistenceException e) {
+      LOGGER.debug("Failed to fetch site {} for queue", siteId, e);
+      return null;
+    }
+
+    NodeAdapter nodeAdapter;
+    try {
+      nodeAdapter = nodeAdapters.factoryFor(site.getType()).create(site.getUrl());
+    } catch (IllegalArgumentException | AdapterException e) {
+      LOGGER.debug("Error retrieving node adapter factory for site {}", site.getUrl(), e);
+      return null;
+    }
+
+    return nodeAdapter;
+  }
+
+  private void doCreate(
+      final Task task,
+      final Metadata metadata,
+      final NodeAdapter source,
+      final NodeAdapter destination)
+      throws InterruptedException {
+    final String metadataId = task.getId();
+    final String operation = task.getOperation().name();
+
+    try {
```

Comment:
I am not sure it can be since the `AdapterInterruptedException` would be caught by the `catch(AdapterException)`. The only way I can think of is to create a functional interface that declares it throws `InterruptedException` and return some generic type. Then call whatever method throws an `AdapterException` through a new method that receives that functional interface that actually does the call and unwraps the `AdapterInterruptedException`. Something like this:

```
@FunctionalInterface
public interface InterruptableSupplier<T> {
    /**
     * Gets a result.
     *
     * @return a result
     */
    T get() throws InterruptedException;
}

private <T> T doInterruptibly(InterruptibleSupplier<T> s) {
  try {
    return s.get();
  } catch (AdapterInterrutedException e) {
    throw new InterruptedException(e);
  }
}
```
Then wrap one of the call below like this:
```
try {
  if (!doInterruptibly(() -> source.exists(metadata)) {
   ...
} catch (AdapterException e) {
...
}
```

You could even extend the `doInterruptily()` method to receive additional info or a `Runnable` to be called if an `AdapterException` is thrown before throwing back, thus removing the need to put a try/catch altogether here.